drivers/shared/src/vic_log.c

```diff
@@ -29,80 +29,81 @@
  *****************************************************************************/
 
 #include <time.h>
-
 #include <vic_def.h>
 #include <vic_run.h>
 #include <vic_driver_shared.h>
 
 /******************************************************************************
- * @brief    Get string of current date and time.  Format YYMMDD-SSSSS.
+ * @brief    Finalize logging
  *****************************************************************************/
-char*
-get_current_datetime()
+void
+finalize_logging(void)
 {
-    const int size = 20;
-    char     *cdt = (char*)malloc(sizeof(char) * size);
-    char      ymd[8];
-    unsigned  seconds_since_midnight;
+    extern FILE *LOG_DEST;
 
-    if (cdt == NULL) {
-        return NULL;
+    if (!(LOG_DEST == stdout || LOG_DEST == stderr)) {
+        fclose(LOG_DEST);
     }
+}
 
-    memset(cdt, 0, size);
-
-    time_t currDateTime;
-    currDateTime = time(NULL);
+/******************************************************************************
+ * @brief    Get string of current date and time.  Format YYMMDD-SSSSS.
+ *****************************************************************************/
+void
+get_current_datetime(char *cdt)
+{
+    char      ymd[8];
+    struct tm timeinfo;
+    unsigned  seconds_since_midnight;
+    time_t    curr_date_time;
 
-    if (currDateTime == -1) {
-        return NULL;
+    curr_date_time = time(NULL);
+    if (curr_date_time == -1) {
+        return;
     }
 
-    struct tm *timeinfo = localtime(&currDateTime);
+    localtime_r(&curr_date_time, &timeinfo);
 
-    seconds_since_midnight = (unsigned)currDateTime % 86400;
+    seconds_since_midnight = (unsigned) curr_date_time % CONST_CDAY;
 
-    if (strftime(ymd, 7, "%y%m%d", timeinfo) == 0) {
-        return NULL;
+    if (strftime(ymd, 7, "%y%m%d", &timeinfo) == 0) {
+        return;
     }
 
     sprintf(cdt, "%s-%05d", ymd, seconds_since_midnight);
-
-    return cdt;
 }
 
 /******************************************************************************
  * @brief    Make logfile name string.
  *****************************************************************************/
-char*
-get_logname(const char *path)
+void
+get_logname(const char *path,
+            int         id,
+            char       *filename)
 {
-    char *timestamp = get_current_datetime();
-    char *prefix = "vic.log.";
+    char  timestamp[MAXSTRING];
```

Comment:
Predefine buffer sizes for temporary string ([MAXSTRING]) rather than dynamic allocation.